src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -531,5 +542,707 @@ public void Add(NamespaceOrTypeSymbol symbol)
+                        }
+
+                        alias.Alias.CheckConstraints(diagnostics);
+
+                        semanticDiagnostics.AddRange(diagnostics.DiagnosticBag);
+                        recordImportDependencies(alias.UsingDirective!, target);
+                    }
+                }
+
+                var corLibrary = compilation.SourceAssembly.CorLibrary;
+                var conversions = new TypeConversions(corLibrary);
+                foreach (var @using in usingsAndDiagnostics.UsingNamespacesOrTypes)
+                {
+                    diagnostics.Clear();
+                    diagnostics.AddDependencies(@using.Dependencies);
+
+                    NamespaceOrTypeSymbol target = @using.NamespaceOrType;
+
+                    // Check if `using static` directives meet constraints.
+                    UsingDirectiveSyntax usingDirective = @using.UsingDirective!;
+                    if (target.IsType)
+                    {
+                        var typeSymbol = (TypeSymbol)target;
+                        var location = usingDirective.Name.Location;
+                        typeSymbol.CheckAllConstraints(compilation, conversions, location, diagnostics);
+                    }
+
+                    semanticDiagnostics.AddRange(diagnostics.DiagnosticBag);
+                    recordImportDependencies(usingDirective, target);
+                }
+
+                // Force resolution of extern aliases.
+                foreach (var alias in externAliasesAndDiagnostics.ExternAliases)
+                {
+                    if (alias.SkipInLookup)
+                    {
+                        continue;
+                    }
+
+                    var target = (NamespaceSymbol)alias.Alias.GetAliasTarget(null);
+                    Debug.Assert(target.IsGlobalNamespace);
+
+                    if (alias.Alias is AliasSymbolFromSyntax aliasFromSyntax)
+                    {
+                        semanticDiagnostics.AddRange(aliasFromSyntax.AliasTargetDiagnostics.DiagnosticBag!);
+                    }
+
+                    if (!Compilation.ReportUnusedImportsInTree(alias.ExternAliasDirective!.SyntaxTree))
+                    {
+                        diagnostics.Clear();
+                        diagnostics.AddAssembliesUsedByNamespaceReference(target);
+                        compilation.AddUsedAssemblies(diagnostics.DependenciesBag);
+                    }
+                }
+
+                semanticDiagnostics.AddRange(externAliasesAndDiagnostics.Diagnostics);
+
+                if (usingsAndDiagnostics.Diagnostics?.IsEmptyWithoutResolution == false)
+                {
+                    semanticDiagnostics.AddRange(usingsAndDiagnostics.Diagnostics.AsEnumerable());
+                }
+
+                diagnostics.Free();
+
+                void recordImportDependencies(UsingDirectiveSyntax usingDirective, NamespaceOrTypeSymbol target)
```

Comment:
>recordImportDependencies [](start = 21, length = 24)

nit: consider adjusting to make the local function static to avoid closure allocation (pass `Compilation` and `diagnostics` in)

Reply:
> consider adjusting to make the local function static to avoid closure allocation (pass Compilation and diagnostics in)

I thought compiler could do exactly that at rewrite.

---
In reply to: [588799806](https://github.com/dotnet/roslyn/pull/51676#discussion_r588799806) [](ancestors = 588799806)